src/agent/src/initdata.rs

```diff
@@ -0,0 +1,232 @@
+}
+
+pub async fn initialize_initdata(logger: &Logger) -> Result<Option<Vec<u8>>> {
+    let logger = logger.new(o!("subsystem" => "initdata"));
+    let candidate_initdata_devices = detect_candidate_initdata_devices(&logger).await?;
+    if candidate_initdata_devices.is_empty() {
+        info!(
+            logger,
+            "Initdata device not found, skip initdata initialization"
+        );
+        return Ok(None);
+    };
+
+    tokio::fs::create_dir_all(INITDATA_PATH)
+        .await
+        .inspect_err(|e| error!(logger, "Failed to create initdata dir: {e:?}"))?;
+
+    let initdata_plaintext_option = candidate_initdata_devices
+        .iter()
+        .filter_map(|device_path| match try_read_initdata(device_path) {
+            Ok(r) => r,
+            Err(e) => {
+                debug!(
+                    logger,
+                    "Failed to read initdata from device {device_path}: {e:?}"
+                );
+                None
+            }
+        })
+        .next();
+
+    let Some(initdata_content) = initdata_plaintext_option else {
+        info!(
+            logger,
+            "Initdata device not found, skip initdata initialization"
+        );
+        return Ok(None);
+    };
+
+    let initdata_plaintext = String::from_utf8(initdata_content.clone())?;
+    debug!(logger, "Initdata full content: {initdata_plaintext}");
+
+    let initdata: Initdata =
+        toml::from_slice(&initdata_content).context("parse initdata failed")?;
+    info!(logger, "Initdata version: {}", initdata.version);
+
+    if initdata.version != INITDATA_VERSION {
+        bail!("Unsupported initdata version");
+    }
+
+    let digest = match &initdata.algorithm[..] {
+        "sha256" => {
+            let mut hasher = Sha256::new();
+            hasher.update(&initdata_content);
+            hasher.finalize().to_vec()
+        }
+        "sha384" => {
+            let mut hasher = Sha384::new();
+            hasher.update(&initdata_content);
+            hasher.finalize().to_vec()
+        }
+        "sha512" => {
+            let mut hasher = Sha512::new();
+            hasher.update(&initdata_content);
+            hasher.finalize().to_vec()
+        }
+        others => bail!("Unsupported hash algorithm {others}"),
+    };
+
+    if let Some(config) = initdata.data.aa_config {
+        tokio::fs::write(AA_CONFIG_PATH, config)
+            .await
+            .context("write aa config failed")?;
+        info!(logger, "write AA config from initdata");
+    }
+
+    if let Some(config) = initdata.data.cdh_config {
+        tokio::fs::write(CDH_CONFIG_PATH, config)
+            .await
+            .context("write cdh config failed")?;
+        info!(logger, "write CDH config from initdata");
+    }
+
+    if let Some(_policy) = initdata.data.policy {
+        #[cfg(feature = "agent-policy")]
+        {
+            let mut policy_agent = crate::AGENT_POLICY.lock().await;
+            policy_agent.set_policy(&_policy).await?;
```

Comment:
This is a bit problematic: the policy is already active, but we did not verify the digest yet. Even if the agent service is not running yet when this is called, I'd rather do it the other way 'round: verify, then set. Could we just write the policy to file here and then load it after starting the AA?

Reply:
Sounds great. Let me do a little change based on your opinion.